# Data Assignment 1 - `t54zheng` (20939203)


In [5]:
import pandas as pd
from pandasql import sqldf

## Part 1 - Downloading and Reading Raw Data

WRDS Queries
- [8076602 - crsp](https://wrds-www.wharton.upenn.edu/query-manager/query/8076602/)
- [8076622 - compustat](https://wrds-www.wharton.upenn.edu/query-manager/query/8076622/)


In [2]:
# import raw data
crsp = pd.read_sas("crsp_nasdaq_100.sas7bdat", encoding = 'latin-1')
compustat = pd.read_sas("compustat_nasdaq_100.sas7bdat", encoding = 'latin-1')

In [30]:
crsp

PERMNO       DATE     CUSIP          COMNAM TICKER         PRC  \
0      10107.0 1997-01-31  59491810  MICROSOFT CORP   MSFT  102.000000   
1      10107.0 1997-02-28  59491810  MICROSOFT CORP   MSFT   97.500000   
2      10107.0 1997-03-31  59491810  MICROSOFT CORP   MSFT   91.687500   
3      10107.0 1997-04-30  59491810  MICROSOFT CORP   MSFT  121.500000   
4      10107.0 1997-05-30  59491810  MICROSOFT CORP   MSFT  124.000000   
...        ...        ...       ...             ...    ...         ...   
23992  93436.0 2022-08-31  88160R10       TESLA INC   TSLA  275.609985   
23993  93436.0 2022-09-30  88160R10       TESLA INC   TSLA  265.250000   
23994  93436.0 2022-10-31  88160R10       TESLA INC   TSLA  227.539993   
23995  93436.0 2022-11-30  88160R10       TESLA INC   TSLA  194.699997   
23996  93436.0 2022-12-30  88160R10       TESLA INC   TSLA  123.180000   

            RET      RETX     SHROUT    VWRETD    SPRTRN  YEAR  MONTH  monthid  
0      0.234493  0.234493  1198000.0  0.053473  0.061317  1997      1       13  
1     -0.044118 -0.044118  1198000.0 -0.001067  0.005928  1997      2       14  
2     -0.059615 -0.059615  1191000.0 -0.044889 -0.042614  1997      3       15  
3      0.325153  0.325153  1191000.0  0.042396  0.058406  1997      4       16  
4      0.020576  0.020576  1191000.0  0.071640  0.058577  1997      5       17  
...         ...       ...        ...       ...       ...   ...    ...      ...  
23992 -0.072489 -0.072489  3133470.0 -0.036233 -0.042440  2022      8      320  
23993 -0.037589 -0.037589  3158000.0 -0.091323 -0.093396  2022      9      321  
23994 -0.142168 -0.142168  3157752.0  0.077394  0.079863  2022     10      322  
23995 -0.144326 -0.144326  3157752.0  0.052354  0.053753  2022     11      323  
23996 -0.367334 -0.367334  3164000.0 -0.057119 -0.058971  2022     12      324  

[23997 rows x 14 columns]

In [31]:
set(crsp.loc[crsp.PERMNO == 93436.0].TICKER)

{'TSLA', nan}

In [4]:
compustat

GVKEY   DATADATE  FYEARQ  FQTR                    CONM   TIC DATACQTR  \
0     001161 1996-03-31  1996.0   1.0  ADVANCED MICRO DEVICES   AMD   1996Q1   
1     001161 1996-06-30  1996.0   2.0  ADVANCED MICRO DEVICES   AMD   1996Q2   
2     001161 1996-09-30  1996.0   3.0  ADVANCED MICRO DEVICES   AMD   1996Q3   
3     001161 1996-12-31  1996.0   4.0  ADVANCED MICRO DEVICES   AMD   1996Q4   
4     001161 1997-03-31  1997.0   1.0  ADVANCED MICRO DEVICES   AMD   1997Q1   
...      ...        ...     ...   ...                     ...   ...      ...   
8858  184996 2021-12-31  2021.0   4.0               TESLA INC  TSLA   2021Q4   
8859  184996 2022-03-31  2022.0   1.0               TESLA INC  TSLA   2022Q1   
8860  184996 2022-06-30  2022.0   2.0               TESLA INC  TSLA   2022Q2   
8861  184996 2022-09-30  2022.0   3.0               TESLA INC  TSLA   2022Q3   
8862  184996 2022-12-31  2022.0   4.0               TESLA INC  TSLA   2022Q4   

     DATAFQTR        ATQ       CEQQ  EPSPXQ       IBQ      SALEQ CONSOL  \
0      1996Q1   2954.809   2121.948    0.18    25.327    544.212      C   
1      1996Q2   2845.025   2071.355   -0.26   -34.672    455.077      C   
2      1996Q3   3053.992   2036.658   -0.28   -38.362    456.862      C   
3      1996Q4   3145.283   2021.878   -0.15   -21.243    496.868      C   
4      1997Q1   3426.126   2043.757    0.09    12.951    551.999      C   
...       ...        ...        ...     ...       ...        ...    ...   
8858   2021Q4  62131.000  30189.000    2.29  2321.000  17719.000      C   
8859   2022Q1  66038.000  34085.000    3.20  3318.000  18756.000      C   
8860   2022Q2  68513.000  36376.000    2.18  2259.000  16934.000      C   
8861   2022Q3  74426.000  39851.000    1.05  3292.000  21454.000      C   
8862   2022Q4  82338.000  44704.000    1.18  3687.000  24318.000      C   

     INDFMT DATAFMT POPSRC CURCDQ COSTAT  
0      INDL     STD      D    USD      A  
1      INDL     STD      D    USD      A  
2      INDL     STD      D    USD      A  
3      INDL     STD      D    USD      A  
4      INDL     STD      D    USD      A  
...     ...     ...    ...    ...    ...  
8858   INDL     STD      D    USD      A  
8859   INDL     STD      D    USD      A  
8860   INDL     STD      D    USD      A  
8861   INDL     STD      D    USD      A  
8862   INDL     STD      D    USD      A  

[8863 rows x 19 columns]

### Data Cleanup

- Notice that there are more tickers in the crsp data than there are companies
- This means that some companies change ticker over the observed time period
- this means that we should actually include sort stocks by permno (as suggested in the assignment)

In [25]:
# Demonstrate this value is greater than 99 (size of our nasdaq 100 data)
print(f'CRSP Tickers - {len(set(crsp.TICKER))}')
print(f'Compustat Tickers - {len(set(compustat.TIC))}')

CRSP Tickers - 119
Compustat Tickers - 99


In [68]:
# To fix this, we will just add permnos to our compustat data
translation_df = pd.read_excel("nasdaq100_pop.xlsx", sheet_name="all", converters={"GVKEY":str, "PERMNO":str})
translation_dict = translation_df[["GVKEY", "PERMNO"]].set_index("GVKEY").to_dict()["PERMNO"]

In [67]:
compustat['PERMNO'] = compustat.apply(lambda row: translation_dict[row.GVKEY], axis=1)
compustat

GVKEY   DATADATE  FYEARQ  FQTR                    CONM   TIC DATACQTR  \
0     001161 1996-03-31  1996.0   1.0  ADVANCED MICRO DEVICES   AMD   1996Q1   
1     001161 1996-06-30  1996.0   2.0  ADVANCED MICRO DEVICES   AMD   1996Q2   
2     001161 1996-09-30  1996.0   3.0  ADVANCED MICRO DEVICES   AMD   1996Q3   
3     001161 1996-12-31  1996.0   4.0  ADVANCED MICRO DEVICES   AMD   1996Q4   
4     001161 1997-03-31  1997.0   1.0  ADVANCED MICRO DEVICES   AMD   1997Q1   
...      ...        ...     ...   ...                     ...   ...      ...   
8858  184996 2021-12-31  2021.0   4.0               TESLA INC  TSLA   2021Q4   
8859  184996 2022-03-31  2022.0   1.0               TESLA INC  TSLA   2022Q1   
8860  184996 2022-06-30  2022.0   2.0               TESLA INC  TSLA   2022Q2   
8861  184996 2022-09-30  2022.0   3.0               TESLA INC  TSLA   2022Q3   
8862  184996 2022-12-31  2022.0   4.0               TESLA INC  TSLA   2022Q4   

     DATAFQTR        ATQ       CEQQ  ...  CONSOL  INDFMT  DATAFMT POPSRC  \
0      1996Q1   2954.809   2121.948  ...       C    INDL      STD      D   
1      1996Q2   2845.025   2071.355  ...       C    INDL      STD      D   
2      1996Q3   3053.992   2036.658  ...       C    INDL      STD      D   
3      1996Q4   3145.283   2021.878  ...       C    INDL      STD      D   
4      1997Q1   3426.126   2043.757  ...       C    INDL      STD      D   
...       ...        ...        ...  ...     ...     ...      ...    ...   
8858   2021Q4  62131.000  30189.000  ...       C    INDL      STD      D   
8859   2022Q1  66038.000  34085.000  ...       C    INDL      STD      D   
8860   2022Q2  68513.000  36376.000  ...       C    INDL      STD      D   
8861   2022Q3  74426.000  39851.000  ...       C    INDL      STD      D   
8862   2022Q4  82338.000  44704.000  ...       C    INDL      STD      D   

     CURCDQ COSTAT  YEAR MONTH monthid  PERMNO  
0       USD      A  1996     3       3   61241  
1       USD      A  1996     6       6   61241  
2       USD      A  1996     9       9   61241  
3       USD      A  1996    12      12   61241  
4       USD      A  1997     3      15   61241  
...     ...    ...   ...   ...     ...     ...  
8858    USD      A  2021    12     312   93436  
8859    USD      A  2022     3     315   93436  
8860    USD      A  2022     6     318   93436  
8861    USD      A  2022     9     321   93436  
8862    USD      A  2022    12     324   93436  

[8863 rows x 23 columns]

## Part 2 - Merging the data

In [10]:
# Enumerate data using monthid
crsp['YEAR'] = pd.DatetimeIndex(crsp['DATE']).year
crsp['MONTH'] = pd.DatetimeIndex(crsp['DATE']).month
crsp['monthid'] = ((crsp['YEAR']) - 1996) * 12 + crsp['MONTH']
sqldf("SELECT * FROM crsp LIMIT 5")

PERMNO                        DATE     CUSIP          COMNAM TICKER  \
0  10107.0  1997-01-31 00:00:00.000000  59491810  MICROSOFT CORP   MSFT   
1  10107.0  1997-02-28 00:00:00.000000  59491810  MICROSOFT CORP   MSFT   
2  10107.0  1997-03-31 00:00:00.000000  59491810  MICROSOFT CORP   MSFT   
3  10107.0  1997-04-30 00:00:00.000000  59491810  MICROSOFT CORP   MSFT   
4  10107.0  1997-05-30 00:00:00.000000  59491810  MICROSOFT CORP   MSFT   

        PRC       RET      RETX     SHROUT    VWRETD    SPRTRN  YEAR  MONTH  \
0  102.0000  0.234493  0.234493  1198000.0  0.053473  0.061317  1997      1   
1   97.5000 -0.044118 -0.044118  1198000.0 -0.001067  0.005928  1997      2   
2   91.6875 -0.059615 -0.059615  1191000.0 -0.044889 -0.042614  1997      3   
3  121.5000  0.325153  0.325153  1191000.0  0.042396  0.058406  1997      4   
4  124.0000  0.020576  0.020576  1191000.0  0.071640  0.058577  1997      5   

   monthid  
0       13  
1       14  
2       15  
3       16  
4       17

In [11]:
compustat['YEAR'] = pd.DatetimeIndex(compustat['DATADATE']).year
compustat['MONTH'] = pd.DatetimeIndex(compustat['DATADATE']).month
compustat['monthid'] = ((compustat['YEAR']) - 1996) * 12 + compustat['MONTH']
sqldf("SELECT * FROM compustat LIMIT 5")

GVKEY                    DATADATE  FYEARQ  FQTR                    CONM  \
0  001161  1996-03-31 00:00:00.000000  1996.0   1.0  ADVANCED MICRO DEVICES   
1  001161  1996-06-30 00:00:00.000000  1996.0   2.0  ADVANCED MICRO DEVICES   
2  001161  1996-09-30 00:00:00.000000  1996.0   3.0  ADVANCED MICRO DEVICES   
3  001161  1996-12-31 00:00:00.000000  1996.0   4.0  ADVANCED MICRO DEVICES   
4  001161  1997-03-31 00:00:00.000000  1997.0   1.0  ADVANCED MICRO DEVICES   

   TIC DATACQTR DATAFQTR       ATQ      CEQQ  ...    SALEQ  CONSOL  INDFMT  \
0  AMD   1996Q1   1996Q1  2954.809  2121.948  ...  544.212       C    INDL   
1  AMD   1996Q2   1996Q2  2845.025  2071.355  ...  455.077       C    INDL   
2  AMD   1996Q3   1996Q3  3053.992  2036.658  ...  456.862       C    INDL   
3  AMD   1996Q4   1996Q4  3145.283  2021.878  ...  496.868       C    INDL   
4  AMD   1997Q1   1997Q1  3426.126  2043.757  ...  551.999       C    INDL   

  DATAFMT POPSRC CURCDQ COSTAT  YEAR MONTH  monthid  
0     STD      D    USD      A  1996     3        3  
1     STD      D    USD      A  1996     6        6  
2     STD      D    USD      A  1996     9        9  
3     STD      D    USD      A  1996    12       12  
4     STD      D    USD      A  1997     3       15  

[5 rows x 22 columns]

In [69]:
df_compustat_crsp = sqldf("SELECT a.*, b.* \
                           FROM crsp as a \
                           LEFT JOIN compustat as b \
                           ON a.PERMNO = b.PERMNO and a.monthid >= b.monthid + 4 and a.monthid <= b.monthid + 6") # 3-month lag

In [100]:
df_compustat_crsp

PERMNO                        DATE     CUSIP          COMNAM TICKER  \
0      10107.0  1997-01-31 00:00:00.000000  59491810  MICROSOFT CORP   MSFT   
1      10107.0  1997-02-28 00:00:00.000000  59491810  MICROSOFT CORP   MSFT   
2      10107.0  1997-03-31 00:00:00.000000  59491810  MICROSOFT CORP   MSFT   
3      10107.0  1997-04-30 00:00:00.000000  59491810  MICROSOFT CORP   MSFT   
4      10107.0  1997-05-30 00:00:00.000000  59491810  MICROSOFT CORP   MSFT   
...        ...                         ...       ...             ...    ...   
23998  93436.0  2022-08-31 00:00:00.000000  88160R10       TESLA INC   TSLA   
23999  93436.0  2022-09-30 00:00:00.000000  88160R10       TESLA INC   TSLA   
24000  93436.0  2022-10-31 00:00:00.000000  88160R10       TESLA INC   TSLA   
24001  93436.0  2022-11-30 00:00:00.000000  88160R10       TESLA INC   TSLA   
24002  93436.0  2022-12-30 00:00:00.000000  88160R10       TESLA INC   TSLA   

              PRC       RET      RETX     SHROUT    VWRETD  ...  CONSOL  \
0      102.000000  0.234493  0.234493  1198000.0  0.053473  ...       C   
1       97.500000 -0.044118 -0.044118  1198000.0 -0.001067  ...       C   
2       91.687500 -0.059615 -0.059615  1191000.0 -0.044889  ...       C   
3      121.500000  0.325153  0.325153  1191000.0  0.042396  ...       C   
4      124.000000  0.020576  0.020576  1191000.0  0.071640  ...       C   
...           ...       ...       ...        ...       ...  ...     ...   
23998  275.609985 -0.072489 -0.072489  3133470.0 -0.036233  ...       C   
23999  265.250000 -0.037589 -0.037589  3158000.0 -0.091323  ...       C   
24000  227.539993 -0.142168 -0.142168  3157752.0  0.077394  ...       C   
24001  194.699997 -0.144326 -0.144326  3157752.0  0.052354  ...       C   
24002  123.180000 -0.367334 -0.367334  3164000.0 -0.057119  ...       C   

       INDFMT  DATAFMT  POPSRC CURCDQ COSTAT    YEAR  MONTH monthid PERMNO  
0        INDL      STD       D    USD      A  1996.0    9.0     9.0  10107  
1        INDL      STD       D    USD      A  1996.0    9.0     9.0  10107  
2        INDL      STD       D    USD      A  1996.0    9.0     9.0  10107  
3        INDL      STD       D    USD      A  1996.0   12.0    12.0  10107  
4        INDL      STD       D    USD      A  1996.0   12.0    12.0  10107  
...       ...      ...     ...    ...    ...     ...    ...     ...    ...  
23998    INDL      STD       D    USD      A  2022.0    3.0   315.0  93436  
23999    INDL      STD       D    USD      A  2022.0    3.0   315.0  93436  
24000    INDL      STD       D    USD      A  2022.0    6.0   318.0  93436  
24001    INDL      STD       D    USD      A  2022.0    6.0   318.0  93436  
24002    INDL      STD       D    USD      A  2022.0    6.0   318.0  93436  

[23997 rows x 37 columns]

## Notice Duplicates
- Notice that our dataframe size exceeds our initial data size from CRSP of 23997 rows, so we have some duplicates
- And indeed we do (I found them using the command below, but showing them first so it's easier to understand)

In [97]:
df_compustat_crsp[6870:6881]

PERMNO                        DATE     CUSIP                  COMNAM  \
6870  45911.0  2002-04-30 00:00:00.000000  83088M10    ALPHA INDUSTRIES INC   
6871  45911.0  2002-05-31 00:00:00.000000  83088M10    ALPHA INDUSTRIES INC   
6872  45911.0  2002-05-31 00:00:00.000000  83088M10    ALPHA INDUSTRIES INC   
6873  45911.0  2002-06-28 00:00:00.000000  83088M10  SKYWORKS SOLUTIONS INC   
6874  45911.0  2002-06-28 00:00:00.000000  83088M10  SKYWORKS SOLUTIONS INC   
6875  45911.0  2002-07-31 00:00:00.000000  83088M10  SKYWORKS SOLUTIONS INC   
6876  45911.0  2002-07-31 00:00:00.000000  83088M10  SKYWORKS SOLUTIONS INC   
6877  45911.0  2002-08-30 00:00:00.000000  83088M10  SKYWORKS SOLUTIONS INC   
6878  45911.0  2002-08-30 00:00:00.000000  83088M10  SKYWORKS SOLUTIONS INC   
6879  45911.0  2002-09-30 00:00:00.000000  83088M10  SKYWORKS SOLUTIONS INC   
6880  45911.0  2002-09-30 00:00:00.000000  83088M10  SKYWORKS SOLUTIONS INC   

     TICKER    PRC       RET      RETX    SHROUT    VWRETD  ...  CONSOL  \
6870   AHAA  12.25 -0.196721 -0.196721   44260.0 -0.049600  ...       C   
6871   AHAA  10.19 -0.168163 -0.168163   44291.0 -0.010510  ...       C   
6872   AHAA  10.19 -0.168163 -0.168163   44291.0 -0.010510  ...       C   
6873   SWKS   5.55 -0.455348 -0.455348  137368.0 -0.070259  ...       C   
6874   SWKS   5.55 -0.455348 -0.455348  137368.0 -0.070259  ...       C   
6875   SWKS   2.90 -0.477477 -0.477477  137510.0 -0.081125  ...       C   
6876   SWKS   2.90 -0.477477 -0.477477  137510.0 -0.081125  ...       C   
6877   SWKS   4.20  0.448276  0.448276  137510.0  0.007949  ...       C   
6878   SWKS   4.20  0.448276  0.448276  137510.0  0.007949  ...       C   
6879   SWKS   4.53  0.078572  0.078572  137589.0 -0.099923  ...       C   
6880   SWKS   4.53  0.078572  0.078572  137589.0 -0.099923  ...       C   

      INDFMT  DATAFMT  POPSRC CURCDQ COSTAT    YEAR  MONTH monthid PERMNO  
6870    INDL      STD       D    USD      A  2001.0   12.0    72.0  45911  
6871    INDL      STD       D    USD      A  2001.0   12.0    72.0  45911  
6872    INDL      STD       D    USD      A  2001.0   12.0    72.0  45911  
6873    INDL      STD       D    USD      A  2001.0   12.0    72.0  45911  
6874    INDL      STD       D    USD      A  2001.0   12.0    72.0  45911  
6875    INDL      STD       D    USD      A  2002.0    3.0    75.0  45911  
6876    INDL      STD       D    USD      A  2002.0    3.0    75.0  45911  
6877    INDL      STD       D    USD      A  2002.0    3.0    75.0  45911  
6878    INDL      STD       D    USD      A  2002.0    3.0    75.0  45911  
6879    INDL      STD       D    USD      A  2002.0    3.0    75.0  45911  
6880    INDL      STD       D    USD      A  2002.0    3.0    75.0  45911  

[11 rows x 37 columns]

In [91]:
# These are our dropped rows
df_compustat_crsp[~df_compustat_crsp.index.isin(df_compustat_crsp.drop_duplicates(["PERMNO", "DATE"], ignore_index=False).index)]

PERMNO                        DATE     CUSIP                  COMNAM  \
6870  45911.0  2002-04-30 00:00:00.000000  83088M10    ALPHA INDUSTRIES INC   
6872  45911.0  2002-05-31 00:00:00.000000  83088M10    ALPHA INDUSTRIES INC   
6874  45911.0  2002-06-28 00:00:00.000000  83088M10  SKYWORKS SOLUTIONS INC   
6876  45911.0  2002-07-31 00:00:00.000000  83088M10  SKYWORKS SOLUTIONS INC   
6878  45911.0  2002-08-30 00:00:00.000000  83088M10  SKYWORKS SOLUTIONS INC   
6880  45911.0  2002-09-30 00:00:00.000000  83088M10  SKYWORKS SOLUTIONS INC   

     TICKER    PRC       RET      RETX    SHROUT    VWRETD  ...  CONSOL  \
6870   AHAA  12.25 -0.196721 -0.196721   44260.0 -0.049600  ...       C   
6872   AHAA  10.19 -0.168163 -0.168163   44291.0 -0.010510  ...       C   
6874   SWKS   5.55 -0.455348 -0.455348  137368.0 -0.070259  ...       C   
6876   SWKS   2.90 -0.477477 -0.477477  137510.0 -0.081125  ...       C   
6878   SWKS   4.20  0.448276  0.448276  137510.0  0.007949  ...       C   
6880   SWKS   4.53  0.078572  0.078572  137589.0 -0.099923  ...       C   

      INDFMT  DATAFMT  POPSRC CURCDQ COSTAT    YEAR  MONTH monthid PERMNO  
6870    INDL      STD       D    USD      A  2001.0   12.0    72.0  45911  
6872    INDL      STD       D    USD      A  2001.0   12.0    72.0  45911  
6874    INDL      STD       D    USD      A  2001.0   12.0    72.0  45911  
6876    INDL      STD       D    USD      A  2002.0    3.0    75.0  45911  
6878    INDL      STD       D    USD      A  2002.0    3.0    75.0  45911  
6880    INDL      STD       D    USD      A  2002.0    3.0    75.0  45911  

[6 rows x 37 columns]

In [101]:
# So we drop the duplicates and we are good
df_compustat_crsp.drop_duplicates(["PERMNO", "DATE"], inplace=True)
df_compustat_crsp

PERMNO                        DATE     CUSIP          COMNAM TICKER  \
0      10107.0  1997-01-31 00:00:00.000000  59491810  MICROSOFT CORP   MSFT   
1      10107.0  1997-02-28 00:00:00.000000  59491810  MICROSOFT CORP   MSFT   
2      10107.0  1997-03-31 00:00:00.000000  59491810  MICROSOFT CORP   MSFT   
3      10107.0  1997-04-30 00:00:00.000000  59491810  MICROSOFT CORP   MSFT   
4      10107.0  1997-05-30 00:00:00.000000  59491810  MICROSOFT CORP   MSFT   
...        ...                         ...       ...             ...    ...   
23998  93436.0  2022-08-31 00:00:00.000000  88160R10       TESLA INC   TSLA   
23999  93436.0  2022-09-30 00:00:00.000000  88160R10       TESLA INC   TSLA   
24000  93436.0  2022-10-31 00:00:00.000000  88160R10       TESLA INC   TSLA   
24001  93436.0  2022-11-30 00:00:00.000000  88160R10       TESLA INC   TSLA   
24002  93436.0  2022-12-30 00:00:00.000000  88160R10       TESLA INC   TSLA   

              PRC       RET      RETX     SHROUT    VWRETD  ...  CONSOL  \
0      102.000000  0.234493  0.234493  1198000.0  0.053473  ...       C   
1       97.500000 -0.044118 -0.044118  1198000.0 -0.001067  ...       C   
2       91.687500 -0.059615 -0.059615  1191000.0 -0.044889  ...       C   
3      121.500000  0.325153  0.325153  1191000.0  0.042396  ...       C   
4      124.000000  0.020576  0.020576  1191000.0  0.071640  ...       C   
...           ...       ...       ...        ...       ...  ...     ...   
23998  275.609985 -0.072489 -0.072489  3133470.0 -0.036233  ...       C   
23999  265.250000 -0.037589 -0.037589  3158000.0 -0.091323  ...       C   
24000  227.539993 -0.142168 -0.142168  3157752.0  0.077394  ...       C   
24001  194.699997 -0.144326 -0.144326  3157752.0  0.052354  ...       C   
24002  123.180000 -0.367334 -0.367334  3164000.0 -0.057119  ...       C   

       INDFMT  DATAFMT  POPSRC CURCDQ COSTAT    YEAR  MONTH monthid PERMNO  
0        INDL      STD       D    USD      A  1996.0    9.0     9.0  10107  
1        INDL      STD       D    USD      A  1996.0    9.0     9.0  10107  
2        INDL      STD       D    USD      A  1996.0    9.0     9.0  10107  
3        INDL      STD       D    USD      A  1996.0   12.0    12.0  10107  
4        INDL      STD       D    USD      A  1996.0   12.0    12.0  10107  
...       ...      ...     ...    ...    ...     ...    ...     ...    ...  
23998    INDL      STD       D    USD      A  2022.0    3.0   315.0  93436  
23999    INDL      STD       D    USD      A  2022.0    3.0   315.0  93436  
24000    INDL      STD       D    USD      A  2022.0    6.0   318.0  93436  
24001    INDL      STD       D    USD      A  2022.0    6.0   318.0  93436  
24002    INDL      STD       D    USD      A  2022.0    6.0   318.0  93436  

[23997 rows x 37 columns]

Finally, we keep the cleaned data from 2000-2022 for our next set of questions.

In [106]:
df_compustat_crsp_2000_2022 = df_compustat_crsp.loc[df_compustat_crsp["DATE"] >= "2000-01-01"]

In [107]:
df_compustat_crsp_2000_2022.reset_index(inplace=True)

In [108]:
df_compustat_crsp_2000_2022

index   PERMNO                        DATE     CUSIP          COMNAM  \
0         36  10107.0  2000-01-31 00:00:00.000000  59491810  MICROSOFT CORP   
1         37  10107.0  2000-02-29 00:00:00.000000  59491810  MICROSOFT CORP   
2         38  10107.0  2000-03-31 00:00:00.000000  59491810  MICROSOFT CORP   
3         39  10107.0  2000-04-28 00:00:00.000000  59491810  MICROSOFT CORP   
4         40  10107.0  2000-05-31 00:00:00.000000  59491810  MICROSOFT CORP   
...      ...      ...                         ...       ...             ...   
22015  23998  93436.0  2022-08-31 00:00:00.000000  88160R10       TESLA INC   
22016  23999  93436.0  2022-09-30 00:00:00.000000  88160R10       TESLA INC   
22017  24000  93436.0  2022-10-31 00:00:00.000000  88160R10       TESLA INC   
22018  24001  93436.0  2022-11-30 00:00:00.000000  88160R10       TESLA INC   
22019  24002  93436.0  2022-12-30 00:00:00.000000  88160R10       TESLA INC   

      TICKER         PRC       RET      RETX     SHROUT  ...  CONSOL  INDFMT  \
0       MSFT   97.875000 -0.161670 -0.161670  5160025.0  ...       C    INDL   
1       MSFT   89.375000 -0.086845 -0.086845  5160025.0  ...       C    INDL   
2       MSFT  106.250000  0.188811  0.188811  5242000.0  ...       C    INDL   
3       MSFT   69.750000 -0.343529 -0.343529  5262405.0  ...       C    INDL   
4       MSFT   62.562500 -0.103047 -0.103047  5262405.0  ...       C    INDL   
...      ...         ...       ...       ...        ...  ...     ...     ...   
22015   TSLA  275.609985 -0.072489 -0.072489  3133470.0  ...       C    INDL   
22016   TSLA  265.250000 -0.037589 -0.037589  3158000.0  ...       C    INDL   
22017   TSLA  227.539993 -0.142168 -0.142168  3157752.0  ...       C    INDL   
22018   TSLA  194.699997 -0.144326 -0.144326  3157752.0  ...       C    INDL   
22019   TSLA  123.180000 -0.367334 -0.367334  3164000.0  ...       C    INDL   

       DATAFMT  POPSRC  CURCDQ COSTAT    YEAR  MONTH  monthid PERMNO  
0          STD       D     USD      A  1999.0    9.0     45.0  10107  
1          STD       D     USD      A  1999.0    9.0     45.0  10107  
2          STD       D     USD      A  1999.0    9.0     45.0  10107  
3          STD       D     USD      A  1999.0   12.0     48.0  10107  
4          STD       D     USD      A  1999.0   12.0     48.0  10107  
...        ...     ...     ...    ...     ...    ...      ...    ...  
22015      STD       D     USD      A  2022.0    3.0    315.0  93436  
22016      STD       D     USD      A  2022.0    3.0    315.0  93436  
22017      STD       D     USD      A  2022.0    6.0    318.0  93436  
22018      STD       D     USD      A  2022.0    6.0    318.0  93436  
22019      STD       D     USD      A  2022.0    6.0    318.0  93436  

[22020 rows x 38 columns]